### ISSUE TITLE
Blog post ideas

### ISSUE THREAD

**Issue by andywhite37** [reazen/relude#175]
Here are some ideas for how-to articles/blog-posts/best practices that we should write at some point:

- [ ] General best practices
  - [ ] Push poorly-typed values to the edges of the system
  - [ ] Push side effects to the edges of the system
  - [ ] Passing props compared to magic context/etc.

- [ ] Beginner's guide to functors/applicatives/monads/etc.
  - [ ] High-level descriptions of what they are
  - [ ] Talk about the laws, and why the laws matter (from a practical perspective)
  - [ ] Show how easy it is to implement these things
  - [ ] Talk about how to recognize a monad in the wild
  - [ ] Talk about what to do when you are mixing monads like `Promise` and `Result`, etc.
  - [ ] Talk about how applicatives are for parallel processing and monads are for sequential

- [ ] How `option`, `Result`, and `IO` are related
  - [ ] Maybe as a follow-on to the above beginner's guide

- [ ] Why traversable is so powerful, and how to recognize that pattern
    
- [ ] How to work with JSON in a type-safe and non-exception-throwing way
  - [ ] Best practices for keeping JSON at the edge of the system
  - [ ] Defining model types and composing decoders/encoders

**Comment by mlms13** [comment]
I started working on a demo weather app that highlights more of the _how_ (how to make requests with `relude-fetch`, how to decode the response with `bs-decode`, how to feed all of that into your React reducer and model async state). I haven't published it yet, but I should really get around to polishing that up. It would be awesome to link to articles like these along the way, to explain more of the "why".

I also started writing an article on the module-functor-bs-abstract approach to typeclasses, based on [this comment](https://github.com/reazen/relude/issues/30#issuecomment-464189001), and I've been brainstorming examples to highlight the benefits of getting "functions for free" by combining typeclasses. I think "sum a list of ints" and "join an array of strings" are solid intro examples, and going from there to "you've always wanted traverse/sequence, and now you don't have to write it by hand anymore" is a useful next step.

In conclusion, I'm excited to start tackling some of this.

**Comment by andywhite37** [comment]
FYI - I posted my first blog article if anyone is interested!  This one is about the `FUNCTOR` typeclass.

https://andywhite.xyz/posts/2019-11-01-a-laymans-guide-to-functors-in-reasonml/

Let me know what you think, I plan to do one about Applicatives next, then Monads, then go from there.

**Comment by andywhite37** [comment]
FYI - here's my second blog post about applicatives:

https://andywhite.xyz/posts/2019-11-07-a-laymans-guide-to-applicatives-in-reasonml/

**Comment by mlms13** [comment]
I finally wrote an intro to type classes (what they are, why they're useful, how to implement them in Reason): https://dev.to/mlms13/type-classes-in-reasonml-a-world-of-functions-for-free-2lag

I only focused on a few basic ones (Eq, Ord, Semigroup, Monoid), but I hope to write a second post looking at the wealth of functionality that comes from Foldable (and maybe its friend Traversable).

**Comment by jihchi** [comment]
—-
I would like to thanks all contributors of `relude`, especially for the documentation sites and blog posts by @mlms13 and @andywhite3.  Those resources are really helpful for the developer like me coming from JavaScript / web development world. 

I couldn’t find any better docs / posts which could clearly explain terminologies like  `typeclass`, `typeclass instances`, `IO (side effect)`, etc. . and along with concrete examples.
